src/java/com/sarality/app/view/dialog/DynamicListDialogFragment.java

```diff
@@ -20,29 +19,64 @@
  */
 public abstract class DynamicListDialogFragment<T> extends BaseDialogFragment implements DataConsumer<List<T>> {
 
-  private final DataSource<List<T>> dataSource;
-  private final ListViewAdapter<T> adapter;
-
-  public DynamicListDialogFragment(FragmentActivity activity, ListViewRowRenderer<T> renderer,
-      DataSource<List<T>> dataSource) {
-    super(activity);
-    this.dataSource = dataSource;
-    this.adapter = new ListViewAdapter<T>(activity, dataSource.getData(), renderer);
-  }
+  private ListViewAdapter<T> adapter;
 
   @Override
   public void onActivityCreated(Bundle savedInstanceState) {
     super.onActivityCreated(savedInstanceState);
-    // Start loading data here.
-    getFragmentActivity().getSupportLoaderManager()
-        .initLoader(0, null, new DataLoader<List<T>>(getActivity(), dataSource, this));
+    DataSource<List<T>> dataSource = getDataSource();
+    startLoad(dataSource, getLoaderId(), this);
+    if (dataSource != null) {
+      this.adapter = new ListViewAdapter<T>(getActivity(), dataSource.getData(), getListViewRowRenderer());
+    } else {
+      this.adapter = new ListViewAdapter<T>(getActivity(), null, getListViewRowRenderer());
+    }
+  }
+
+  /**
+   * @return Renderer to be used to render each row of the list displayed in the dialog.
+   */
+  protected abstract ListViewRowRenderer<T> getListViewRowRenderer();
+
+  /**
+   * Returns a unique Id to be used for the loader.
+   *
+   * @return Unique Id for the loader.
+   */
+  protected abstract int getLoaderId();
+
+  /**
+   * @return The DataSource used to load data for the fragment.
+   */
+  protected abstract DataSource<List<T>> getDataSource();
+
+  /**
+   *
+   * @param dataSource DataSource to load the data from
+   * @param loaderId Id to used for the loader
+   */
+  protected <D> void startLoad(DataSource<D> dataSource, int loaderId, DataConsumer<D> consumer) {
```

Comment:
does the loaderId need to be passed in. Seems like it is always called with getLoaderId(). Why not just call getLoaderId directly?